### PR TITLE
User,UserRegistrationRequestDto, UserResponseDto, UserMapper, UserRep…

### PR DESCRIPTION
**Pull request by BohdanS-UA**
…ository, UserService, AuthenticationController were added.

### CODE REVIEW

File: src/main/java/mate/academy/bookstorev2/model/User.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstorev2/service/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserMapper userMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        User user = userMapper.toModel(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/service/UserServiceImpl.java:24]: we need to check if user existsByEmail first

File: src/main/java/mate/academy/bookstorev2/validation/UniqueEmailValidator.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookstorev2.repository.user.UserRepository;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class UniqueEmailValidator implements ConstraintValidator<UniqueEmail, String> {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/validation/UniqueEmailValidator.java]: remove

File: src/main/java/mate/academy/bookstorev2/validation/UniqueEmail.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = UniqueEmailValidator.class)
+@Target({ElementType.FIELD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface UniqueEmail {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/validation/UniqueEmail.java]: remove,  redundant

File: src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import mate.academy.bookstorev2.validation.UniqueEmail;
+
+@Getter
+@Setter
+@FieldMatch(firstString = "password",
+        seconfString = "repeatPassword",
+        message = "The password fields must match")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @UniqueEmail
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java]:
```suggestion
```

If you would like to add custom annotation, use some pattern to check valid email

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

File: src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+@Getter
+@Setter
+@FieldMatch(firstString = "password",
+        seconfString = "repeatPassword",
+        message = "The password fields must match")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    private String password;
+    @NotBlank
+    @Email(message = "Email is not valid",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java:20]: Suggestion , test this pattern if it works correctly. Also, you can create your custom validation like for passwords, but with pattern

File: src/main/java/mate/academy/bookstorev2/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -42,4 +43,15 @@ private String getErrorMessage(ObjectError e) {
         }
         return e.getDefaultMessage();
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException exception,
+                                                              WebRequest webRequest) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp: ", LocalDateTime.now());
+        body.put("status: ", HttpStatus.BAD_REQUEST.value());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        body.put("status: ", HttpStatus.CONFLICT.value());
```

File: src/main/java/mate/academy/bookstorev2/exception/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -42,4 +43,15 @@ private String getErrorMessage(ObjectError e) {
...
+        body.put("errors: ", "Bad Request");
+        body.put("message: ", exception.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/exception/CustomGlobalExceptionHandler.java]: same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+@Getter
+@Setter
+@FieldMatch(firstString = "password",
+        seconfString = "repeatPassword",
+        message = "The password fields must match")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Pattern(regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[#$@!%&*?])[A-Za-z\\d#$@!%&*?]{8,}$")
+    //Min 1 uppercase letter
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/dto/UserRegistrationRequestDto.java]: remove commented code

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**